accepted/future-releases/extension-types/feature-specification.md

```diff
@@ -1179,6 +1179,119 @@ all members of `V1, .. Vk` that are not redeclared by a declaration in _DV_
 can be invoked on a receiver of the type introduced by _DV_.*
 
 
+### Changes to type functions and types
+
+The previous sections have introduced some subtype relationships involving
+extension types. This section adds further relationships where extension
+types occur more indirectly.
+
+*For instance, earlier sections stated that every extension type is a
+subtype of the top types (like `Object?`), and that `implements` introduces
+a subtype relationship for extension types. Moreover:*
+
+Assume that `E` denotes an extension type that takes `k` type arguments.
+Corresponding to the subtype rule about covariance for classes, a rule is
+introduced which makes `E<S1 .. Sk>` a subtype of `E<T1 .. Tk>` if
+`Sj` is a subtype of `Tj` for each `j` in 1 .. k.
+
+*For example, `E<int>` is a subtype of `E<Object>` because `int` is a
+subtype of `Object`, also in the case where `E` is an extension type.*
+
+The type function `futureValueType` is updated as follows:
+
+- `futureValueType(S?) = futureValueType(S)`, for any `S`.
+- `futureValueType(Future<S>) = S`, for any `S`.
+- `futureValueType(FutureOr<S>) = S`, for any `S`.
+- `futureValueType(void) = void`.
+- `futureValueType(dynamic) = dynamic`.
+- `futureValueType(E) = S` if `E` is an extension type
+  that implements `Future<S>`.
+- Otherwise, `futureValueType(S) = Object?`, for any `S`.
+
+The Dart 1 algorithm which is used to compute the standard upper bound of
+two distinct interface types will work in the same way as it does today,
+albeit on a different superinterface graph:
+
+We introduce an interface type `Any` which declares `operator ==`,
+`hashCode`, `runtimeType`, `noSuchMethod`, `toString` with the same
+signatures as `Object` does today, and `Object` as well as `Null` have
+`Any` as a direct superinterface.
+
+`Any` is sealed. *That is, we cannot create additional direct subtypes of
+`Any` This is required for the soundness of the rule below about `Object?`
+being a supertype of `Any`.*
+
+*This basically means that we eliminate the anomaly which is created by
+saying that `Object` declares those members, but `Null` "has them too,
+by magic". It also implies that all paths from the top type via `Object`
+are now one step longer, which is significant during the computation of the
+standard upper bound.*
+
+For subtyping, we specify that every type is a subtype of `void`, of
+`dynamic`, and of `Any`.
+
+Moreover, we specify that `Any` is a subtype of `Object?`. *This implies
```

Comment:
That's just saying that any type is a subtype of `void`, `dynamic`, `Any` and `Object?`.

Maybe we could collapse all those to just:
> * `S` \<: `T` if **TOP**(`T`)

(The current subtyping rule is:
> * **Right Top**: if `T1` is a top type (i.e. `dynamic`, or `void`, or `Object`?) then `T0` \<: `T1`

It should probably use "e.g." instead of "i.e." here, and really use **TOP**(`T1`) as the definition.)

And then we keep making **TOP**(`Object?`) be true, like it is to day, and add **TOP**(`Any`) = true as well.

Reply:
Closing, this text has been deleted.